File: 4 sem/algorithms & data structures/recurrency py + java/chip.java
import java.io.*;

public class chip {
    static int module = 1000000007;

    public static int[] numberOfRoutes(int t, int[][]params) {
        int[] result = new int[t];
        for (int i = 0; i < t; i++) {
            int[] poss = new int[params[i][0]];
            poss[0] = 1;
            int sum = 1;
            for (int j = 1; j < params[i][0]; j++) {
                poss[j] = sum;
                sum = (poss[j] + sum) % module;
                if (j - params[i][1] >= 0) {
                    sum = (sum - poss[j - params[i][1]] + module) % module;
                }
            }
            result[i] = poss[params[i][0]-1];
        }
        return result;
    }

    public static void main(String[] args) throws IOException {
        BufferedReader in = new BufferedReader(new FileReader("input.txt"));
        BufferedWriter out = new BufferedWriter(new FileWriter("output.txt"));
        int t = Integer.parseInt(in.readLine());
        int[][] parameters = new int[t][2];
        int[] answer;
        int i = 0;
        String line;
        String[] str;
        while ((line = in.readLine()) != null) {
            str = line.split(" ");
            parameters[i][0] = Integer.parseInt(str[0]);
            parameters[i][1] = Integer.parseInt(str[1]);
            i++;
        }
        answer = numberOfRoutes(t, parameters);
        for (int j : answer) out.write(Integer.toString(j) + '\n');
        out.close();
    }
}

//Алгоритм решения задачи ‘Фишка’, Алина Крагель, 10 группа, 2 курс, ФПМИ:
//Формально нам необходимо перебрать всевозможные суммы, которыми можно представить длину поля, где уникальность хода будет определяться порядком слагаемых, при этом учитывается ограниченное значение любого слагаемого (<=k), что проверяется во внутреннем цикле.
//Решение – функция numberOfRoutes, возвращающая ответ для каждого набора параметров.
//sum – суммарное количество путей, через которые мы можем пройти от начальной позиции до искомой (в цикле до i), вычисление которого мы начинаем с 1 (тк 1 возможность прохождения есть всегда).

File: 4 sem/algorithms & data structures/recurrency py + java/chip.py
import numpy as np


def ways_num(t, parameters):
    base = 1000000007
    results = np.zeros(t, dtype=np.int64)
    for i in range(t):
        poss = np.zeros(parameters[i][0], dtype=np.int64)
        poss[0] = 1
        between = 1
        for j in range(1, parameters[i][0]):
            poss[j] = between
            between = (poss[j] + between) % base
            if j - parameters[i][1] >= 0:
                between = (between - poss[j - parameters[i][1]] + base) % base
        results[i] = poss[parameters[0][i] - 1]
    return results


poss = open("input.txt")
w = open("output.txt", 'w')
t_condition = int(poss.readline())
parameters_condition = np.zeros((t_condition, 2), dtype=np.int64)
for i in range(t_condition):
    parameters_condition[i] = [int(j) for j in poss.readline().split()]
answer = ways_num(t_condition, parameters_condition)
for i in answer:
    w.write(str(i) + '\n')
